src/gui/qgsmapcanvas.cpp

```diff
@@ -604,8 +604,10 @@ void QgsMapCanvas::stopRendering()
   {
     QgsDebugMsg( "CANVAS stop rendering!" );
     mJobCanceled = true;
-    mJob->cancel();
-    Q_ASSERT( !mJob ); // no need to delete here: already deleted in finished()
+    disconnect( mJob, &QgsMapRendererJob::finished, this, &QgsMapCanvas::rendererJobFinished );
+    connect( mJob, &QgsMapRendererQImageJob::finished, mJob, &QgsMapRendererQImageJob::deleteLater );
+    mJob->cancelWithoutBlocking();
+    mJob = nullptr;
```

Comment:
It shouldn't do, because the only signal which is connected after it's cancelled is the one which deletes itself - ie nothing which depends on canvas. I'll add a test for this in any case just to be sure.

Re cache image - there's logic in place in the renderer not to store cache images for cancelled jobs. Maybe that was added after your attempt?